MaxText/train.py

```diff
@@ -301,16 +324,30 @@ def train_loop(config, state=None):
   Returns:
   """
   ( init_rng, writer, checkpoint_manager, state_mesh_annotations, model,
-  mesh, learning_rate_schedule, data_iterator, state) = setup_train_loop(config)
+  mesh, learning_rate_schedule, data_iterator, eval_data_iterator, state) = setup_train_loop(config)
 
-  functional_train, in_shard, out_shard, static_argnums, donate_argnums = maxtext_utils.get_functional_train_with_signature(
+  # pylint: disable=line-too-long
+  functional_train, in_shard, out_shard, static_argnums_train, donate_argnums_train = maxtext_utils.get_functional_train_with_signature(
```

Comment:
Not good to have a "functional_train_with_signature" that takes "is_train". Happy to discuss solutions or for you to propose one!

Reply:
Thank you @rwitten for the comments.

Instead of fusing train and evaluation in the same `functional_train_with_signature`, I am considering to add a `functional_eval_with_signature` in addition to `functional_train_with_signature`, what do you think?

```
def get_functional_eval_step(train_step, model, config):
  # train and eval step share the same api for simplicity:
  #   train:  new_state, metrics = train_step(state, data, is_train=True)
  #   eval:  _, metrics = train_step(state, data, is_train=False)
  return functools.partial(train_step, model, config, is_train=False)

def get_functional_train_with_signature(train_step, mesh, state_mesh_annotations, model, config):
  ...

# add a new independent method
def get_functional_eval_with_signature(train_step, mesh, state_mesh_annotations, model, config):
  ...

```

Reply:
that is a good idea and an improvement. But you've pushed the problem to:
```
return functools.partial(train_step, model, config, is_train=False)
```

Reply:
Hi, Rafi, thank you for your comment. Currently, I combined train and eval [under the same function](https://github.com/google/maxtext/blob/73665db73eb108a23e845b5d8220672d0afd6d71/MaxText/train.py#L181-L252). `is_train` flag is used to differentiate the branches. This is mainly to reuse loss_fn to ensure a compact code style.  
Looks like you are favoring more towards 2 separate functions instead. How about we do this way:
```
# we can almost reuse 99% of loss_fn for both train_step and eval_step
def loss_fn(params, is_train=True):
    logits, intermediate_outputs = model.apply({'params': params},
                         ...
                         enable_dropout=config.enable_dropout if is_train else False,  # only difference
    ...

# independent functions elsewhere.
def train_step
  ...
functional_train = get_functional_train_with_signature
p_train_step = jax.jit(functional_train)

def eval_step
  ...
functional_eval = get_functional_eval_with_signature
p_eval_step = jax.jit(functional_eval)
```

Reply:
SGTM

Reply:
Done.